File: 5. LinkedLists/9. longest_streak/longest_streak.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

a = Node(9)
b = Node(9)
c = Node(1)
d = Node(9)
e = Node(9)
f = Node(9)

a.next = b
b.next = c
c.next = d
d.next = e
e.next = f

# 9 -> 9 -> 1 -> 9 -> 9 -> 9

def longest_streak(head):

    if head is None:
        return 0

    count = 1
    maxCount = 0

    prev = head
    head = head.next
    while(head is not None):
        if prev.val == head.val:
            count += 1
            prev = head
            head = head.next
        else:
            if(count > maxCount):
                maxCount = count
            count = 1
            prev = head
            head = head.next
    
    if(count > maxCount):
        maxCount = count

    return maxCount

        

print(longest_streak(a))
    
  

File: 5. LinkedLists/9. longest_streak/longest_streak.js

// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// const a = new Node(5);
// const b = new Node(5);
// const c = new Node(7);
// const d = new Node(7);
// const e = new Node(7);
// const f = new Node(6);

// a.next = b;
// b.next = c;
// c.next = d;
// d.next = e;
// e.next = f;

// 5 -> 5 -> 7 -> 7 -> 7 -> 6



// Iterative
const longestStreak = (head) => {
    let current = head;
    let streakCount = 0;
    let maxStreakCount = 0;
    let prev = null;
    while(current !== null){
        if(prev && current.val === prev.val)
            streakCount++;
        else
            streakCount = 1;
        if(streakCount > maxStreakCount)
            maxStreakCount = streakCount;
        prev = current;
        current = current.next; 
    }
    return maxStreakCount;
}

// Complexity
// n = number of nodes
// Time: O(n)
// Space: O(1)


// let ans = longestStreak(a); 
// console.log(`Ans: ${ans}`);